Fix bug where wrong select option would get selected.

mtyaka opened the pull request:
This fixes the bug with select elements reported in #107.

The main problem was looping through children of `toEl` when determining the correct `selectedIndex`, since nodes from `toEl` can get moved to `fromEl` while morphing in some cases, so by the time the select handler is invoked, `toEl` may already be missing some of its original children.

On the other hand, at the time when the select handler is invoked, `fromEl`'s children are already fully patched and in place.

close #128 #107

AutoSponge commented:
This looks like a good fix. But, I haven't tested in IE myself.

MelleB commented:
(Bump)

AndyOGo commented:
It's already 2018 and no updates here, is this repo still maintained?

AutoSponge commented on src/specialElHandlers.js:
```diff
@@ -12,11 +12,14 @@ function syncBooleanAttrProp(fromEl, toEl, name) {
-     */
     OPTION: function(fromEl, toEl) {
+        var parentNode = fromEl.parentNode;
+        if (parentNode && parentNode.nodeName === 'SELECT' && !hasAttributeNS(parentNode, null, 'multiple')) {
```
previous code used `.toUpperCase()` every time the `.nodeName` was retrieved. I think that's a good strategy for backward compatibility.

mtyaka replied:
Done, thanks.

AutoSponge commented on src/specialElHandlers.js:
```diff
@@ -61,10 +64,14 @@ export default {
+            // and appended to / removed from fromEl, so using fromEl here is safe and correct.
+            var curChild = fromEl.firstChild;
             while(curChild) {
                 var nodeName = curChild.nodeName;
```
use `.toUpperCase()`

mtyaka replied:
Thanks, good catch. I added `.toUpperCase()` back in.

AutoSponge commented on test/browser/test.js:
```diff
@@ -873,6 +873,34 @@ describe('morphdom' , function() {
+        expect(el1.selectedIndex).to.equal(1);
+    });
+
+    it('should select first option on select elements when no option explicitly selected', function () {
```
I'm not sure this is expected behavior.

Imaging a scenario where changing one form control changes the options in this select input. If the user had previously selected a now deleted option, the input should be UNSET (selectIndex = -1) so that the user must choose again. However, this code updates the choice to one the user didn't make.

mtyaka replied:
I think I was trying to emulate default browser behavior where if you render a `select` the first option gets selected by default, but I like the idea of forcing `selectedIndex` to `-1` when no option is explicitly selected. I updated the code.

mtyaka commented:
@AutoSponge Thanks for reviewing! I addressed your comments and added a workaround for a bug MS Edge when I noticed the test case was not working correctly in Edge.

This is the jsfiddle with updated files: https://jsfiddle.net/5kmwbp2v/27/

Let me know if you have any other concerns.

mtyaka commented:
Pushed a commit to fix issue with `selectedIndex` when using `<optgroup>`s. Test failures seem unrelated.

mrdziuban commented:
@AutoSponge @patrick-steele-idem is there anything preventing this PR from moving forward? I can confirm that it now also fixes #128.

snewcomer commented:
I'm going to take a look at this tomorrow.  If there was still interest to get the tests passing, that would be great!  Otherwise, I'll open a new branch with your changes 👍

mtyaka commented:
Thanks @snewcomer! Rebasing on top of master fixed the test failures.

snewcomer commented on test/browser/test.js:
```diff
@@ -873,6 +873,58 @@ describe('morphdom' , function() {
+        morphdom(el1, el2);
+
+        expect(el1.selectedIndex).to.equal(-1);
+    });
```
What do you think about adding 1 test for options that are `removed`?

mtyaka commented:
Thanks for taking care of resolving the conflicts and merging this @snewcomer!

snewcomer commented on src/specialElHandlers.js:
```diff
@@ -10,11 +10,28 @@ function syncBooleanAttrProp(fromEl, toEl, name) {
+                // We have to reset select element's selectedIndex to -1, otherwise setting
+                // fromEl.selected using the syncBooleanAttrProp below has no effect.
+                // The correct selectedIndex will be set in the SELECT special handler below.
+                parentNode.selectedIndex = -1;
```
@mtyaka What if the default selectedIndex was `0` instead of `-1`.  0 will set the first of the options, but -1 will just blank it out....

snewcomer replied:
nvmd.  This is a bad idea and could inadvertently cause the user to submit the wrong option.  Ignore me!
